docs/reasonreactcompat.md

```diff
@@ -0,0 +1,222 @@
+
+```
+This expression has type array('a)
+but an expression was expected of type
+ReasonReact.reactElement = React.element
+```
+
+In these cases, it is helpful to wrap `children` with `React.array` inside the `Jsx2` compatibility module, like:
+
+```reason
+/* In SomeButton.re */
+module Jsx2 = {
+  let component = ReasonReact.statelessComponent("SomeButton");
+  let make = children => {
+    let children = React.array(children);
+    ReasonReactCompat.wrapReactForReasonReact(
+      make,
+      makeProps(~children, ()),
+      children,
+    );
+  };
+};
+```
+
+The reason behind those errors is that version 3 of JSX doesn't automatically wrap the `children` passed to an element in an array, like version 2 used to do. Using `React.array` in the `Jsx2` module is a way to provide a consistent behavior for usages of the component in both versions of the JSX transform.
+
+## `wrapReasonReactForReact`: Wrapping a v2 component to be used from a v3 component
+
+Sometimes we might need to make the translation the other way around: wrap a v2 component to be used from a v3 component. For example, if we are using some ReasonReact library that has not been updated yet to be compatible with the latest version.
+
+In these cases, we can use `ReasonReact.wrapReasonReactForReact`.
+
+Let's say we have a v2 component `Text` that needs to be used from a v3 component `Heading`:
+
+```reason
+/* In Text.re */
+let component = ReasonReact.statelessComponent("Text");
+
+let make = (~text, _children) => {
+  ...component,
+  render: _self => <span> {ReasonReact.string(text)} </span>,
+};
+```
+
+We can follow the same approach as above and add a `Jsx3` module to the same file:
+
+```reason
+/* Still in Text.re */
+module Jsx3 = {
+  [@bs.obj] external makeProps: (~text: string, unit) => _ = "";
+  let make =
+    ReasonReactCompat.wrapReasonReactForReact(
+      ~component, (reactProps: {. "text": string}) =>
+      make(~text=reactProps##text, [||])
+    );
+};
+```
+
+### Component with `children`
+
+If the v2 component we want to migrate uses `children` we have to take some extra steps to convert it.
+
+For example, let's say we have a `List` v2 component:
+
+```reason
+/* List.re */
+let component = ReasonReact.statelessComponent("List");
+
+let make = (~visible, children) => {
+  ...component,
+  render: _self => visible ? <div> ...children </div> : ReasonReact.null,
+};
+```
+
+The `Jsx3` compat module will look like:
+
+```reason
+/* Still in List.re */
+module Jsx3 = {
+  [@bs.obj]
+  external makeProps: (~visible: bool, ~children: 'children=?, unit) => _ = "";
+  let make =
+    ReasonReactCompat.wrapReasonReactForReact(
+      ~component,
+      (
+        reactProps: {
+          .
+          "visible": bool,
+          "children": 'children,
+        },
+      ) =>
+      make(
+        ~visible=reactProps##visible,
+        reactProps##children
+        ->Js.Undefined.toOption
+        ->Belt.Option.mapWithDefault([||], c => [|c|]),
```

Comment:
@phated @rickyvetter I never know if we should use `React.Children.toArray` here? Can `c => [|c|]` lead to double-wrapped arrays?

Reply:
I'm not sure.  Looking towards Ricky for advice here.

Reply:
Yeah technically you should use toArray here. If you were to call `<Jsx3> {[|<div />|]} </Jsx3>` It would double wrap. That said the cost of double wrapping isn't huge and React will carry over keys and everything correctly so it will never really cause issues to do this.